runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/retries/impl/StandardRetryTokenBucket.kt

```diff
@@ -0,0 +1,141 @@
+    }
+
+    private suspend fun checkoutCapacity(size: Int): Unit = mutex.withLock {
+        refillCapacity()
+
+        if (size <= capacity) {
+            capacity -= size
+        } else {
+            val extraRequiredCapacity = size - capacity
+            val delayMs = ceil(extraRequiredCapacity.toDouble() / options.refillUnitsPerSecond * MS_PER_S).toLong()
+            delay(delayMs)
+            capacity = 0
+        }
+        lastTimestamp = now()
+    }
+
+    private fun now(): Long = clock.now().epochMilliseconds
+
+    private fun refillCapacity() {
+        val refillMs = now() - lastTimestamp
+        val refillSize = floor(options.refillUnitsPerSecond.toDouble() / MS_PER_S * refillMs).toInt()
+        capacity = min(options.maxCapacity, capacity + refillSize)
+    }
+
+    private suspend fun returnCapacity(size: Int): Unit = mutex.withLock {
+        refillCapacity()
+
+        capacity = min(options.maxCapacity, capacity + size)
+        lastTimestamp = now()
+    }
+
+    /**
+     * A standard implementation of a [RetryToken].
+     * @param returnSize The amount of capacity to return to the bucket on a successful try.
+     */
+    inner class StandardRetryToken(val returnSize: Int) : RetryToken {
+        /**
+         * Completes this token because retrying has been abandoned. This implementation doesn't actually increment any
+         * capacity upon failure...capacity just refills based on time.
+         */
+        override suspend fun notifyFailure() {
+            // Do nothing
+        }
+
+        /**
+         * Completes this token because the previous retry attempt was successful.
+         */
+        override suspend fun notifySuccess() {
+            returnCapacity(returnSize)
+        }
+
+        /**
+         * Completes this token and requests another one because the previous retry attempt was unsuccessful.
+         */
+        override suspend fun scheduleRetry(reason: RetryErrorType): RetryToken {
+            val size = when (reason) {
+                RetryErrorType.Timeout, RetryErrorType.Throttling -> options.timeoutRetryCost
+                else -> options.retryCost
+            }
+            checkoutCapacity(size)
+            return StandardRetryToken(size)
+        }
+    }
+}
+
+/**
+ * The configuration options for a [StandardRetryTokenBucket].
+ * @param maxCapacity The maximum capacity for the bucket.
+ * @param refillUnitsPerSecond The amount of capacity to return per second.
+ * @param retryCost The amount of capacity to decrement for standard retries.
+ * @param timeoutRetryCost The amount of capacity to decrement for timeout or throttling retries.
+ * @param initialTryCost The amount of capacity to decrement for the initial try.
+ * @param initialTrySuccessIncrement The amount of capacity to return if the initial try is successful.
+ */
+data class StandardRetryTokenBucketOptions(
+    val maxCapacity: Int,
+    val refillUnitsPerSecond: Int,
```

Comment:
suggestion

`refillRate`

Reply:
I wanted to be explicit about the units here. `refillRate = 10` doesn't really indicate what unit of thing is being refilled nor how often.

Reply:
yeah I mean you could do `refillRatePerSecond` or be fancy and define a type. Something like:



```kotlin
value class TokenRefillRate(val tokensPerSecond: Int)
```

Not a big deal, my brain just kept saying "theres a name for xyz measured per unit of time...RATE"